applications/FluidDynamicsHydraulicsApplication/custom_utilities/hydraulic_fluid_auxiliary_utilities.cpp

```diff
@@ -302,9 +302,141 @@ void HydraulicFluidAuxiliaryUtilities::SetInletFreeSurface(ModelPart &rModelPart
+            auto rCondition = neighb[c_itr];
+            const array_1d<double, 3> &face_neig_normal = rCondition.GetValue(NORMAL);
+            double cos_angle = 1 / (An * norm_2(face_neig_normal)) * inner_prod(face_normal, face_neig_normal);
+            auto &r_geom_neig = rCondition.GetGeometry();
+            auto edgelist_neig = r_geom_neig.GenerateEdges();
+            array_1d<double, 2> ids_neig;
+            array_1d<double, 2> ids_cond;
+            if (cos_angle < acceptable_cos){
+                for (IndexType edge = 0; edge < edgelist.size(); edge++){
+                    for (IndexType i = 0; i < edgelist[edge].size(); i++){
+                        ids_cond[i] = edgelist[edge][i].Id();
+                    }
+                    std::sort(ids_cond.begin(), ids_cond.end());
+                    for (IndexType edge_neg = 0; edge_neg < edgelist_neig.size(); edge_neg++){
+                        for (IndexType i = 0; i < edgelist_neig[edge_neg].size(); i++){
+                            ids_neig[i] = edgelist_neig[edge_neg][i].Id();
+                        }
+                        std::sort(ids_neig.begin(), ids_neig.end());
+                        if (ids_cond == ids_neig){
+                            for (IndexType i = 0; i < edgelist[edge].size(); i++){
+                                double repeated = edgelist[edge][i].GetValue(AUX_INDEX);
+                                repeated += 1;
+                                edgelist[edge][i].SetValue(AUX_INDEX, repeated);
+                            }
+                        }
+                    }
+                }
+            }
         }
+    }
+    block_for_each(rModelPart.Nodes(),[&](NodeType &rNode){
+        if (rNode.GetValue(AUX_INDEX)*0.5>=3 && rNode.Is(MARKER)){
+            Vector veloctiy = ZeroVector(3);
+            rNode.FastGetSolutionStepValue(VELOCITY) = veloctiy;
+            rNode.Fix(VELOCITY_X);
+            rNode.Fix(VELOCITY_Y);
+            rNode.Fix(VELOCITY_Z);
+            double value= rNode.GetValue(AUX_INDEX);
+            rNode.SetValue(AUX_INDEX,value*0.5);
+        }
+        else{
+            rNode.SetValue(AUX_INDEX,0.0);
+        }
+    });
+}
+
+bool HydraulicFluidAuxiliaryUtilities::MaximumWaterDepthChange(const ModelPart &rModelPart){
+    // make sure the Ids of the Nodes start with one
+    const double &min_Z = block_for_each<MinReduction<double>>(rModelPart.Nodes(), [&](const auto &rNode){ return rNode.Z(); });
+    //TODO: Try to create a min reduction with a pointer to avoid this second loop
+    bool maximum_water_depth_change = false;
+    block_for_each(rModelPart.Nodes(), [&](const Node &rNode){
+        if (std::abs(rNode.Z() - min_Z) < 1.0e-10 && !maximum_water_depth_change){
+            const double inlet_phi = rNode.GetValue(AUX_DISTANCE);
+            const double domain_phi = rNode.FastGetSolutionStepValue(DISTANCE);
+            if (std::abs(domain_phi) > std::abs(inlet_phi)){
```

Comment:
Are you sure you want to check this? Shouldn't you check that the absolute value of the difference is below a threshold?